extension/parquet/reader/variant/variant_value.cpp

```diff
@@ -12,6 +24,601 @@ void VariantValue::AddItem(VariantValue &&val) {
+				    static_cast<uint8_t>(VariantLogicalType::VARCHAR);
+			}
+			auto string_data = primitive.GetValueUnsafe<string_t>();
+			auto string_size = string_data.GetSize();
+			VarintEncode<uint32_t>(string_size, blob_data + offsets.data[row]);
+			offsets.data[row] += GetVarintSize(string_size);
+			memcpy(blob_data + offsets.data[row], string_data.GetData(), string_size);
+			offsets.data[row] += string_size;
+			break;
+		}
+		default:
+			throw InternalException("Encountered unrecognized LogicalType in VariantValue::AnalyzeValue: %s",
+			                        primitive.type().ToString());
+		}
+		offsets.values[row]++;
+		break;
+	}
+	default:
+		throw InternalException("VariantValueType not handled");
+	}
+}
+
+//! Copied and modified from 'to_variant.cpp'
+static void InitializeVariants(VariantConversionOffsets &offsets, Vector &result, SelectionVector &keys_selvec,
+                               idx_t &selvec_size) {
+	auto &keys = VariantVector::GetKeys(result);
+	auto keys_data = ListVector::GetData(keys);
+
+	auto &children = VariantVector::GetChildren(result);
+	auto children_data = ListVector::GetData(children);
+
+	auto &values = VariantVector::GetValues(result);
+	auto values_data = ListVector::GetData(values);
+
+	auto &blob = VariantVector::GetData(result);
+	auto blob_data = FlatVector::GetData<string_t>(blob);
+
+	idx_t children_offset = 0;
+	idx_t values_offset = 0;
+	idx_t keys_offset = 0;
+
+	auto keys_sizes = offsets.keys;
+	auto children_sizes = offsets.children;
+	auto values_sizes = offsets.values;
+	auto blob_sizes = offsets.data;
+
+	auto count = values_sizes.size();
+	for (idx_t i = 0; i < count; i++) {
+		auto &keys_entry = keys_data[i];
+		auto &children_entry = children_data[i];
+		auto &values_entry = values_data[i];
+
+		//! keys
+		keys_entry.length = offsets.keys[i];
+		keys_entry.offset = keys_offset;
+		keys_offset += keys_entry.length;
+
+		//! children
+		children_entry.length = children_sizes[i];
+		children_entry.offset = children_offset;
+		children_offset += children_entry.length;
+
+		//! values
+		values_entry.length = values_sizes[i];
+		values_entry.offset = values_offset;
+		values_offset += values_entry.length;
+
+		//! value
+		blob_data[i] = StringVector::EmptyString(blob, blob_sizes[i]);
+	}
+
+	//! Reserve for the children of the lists
+	ListVector::Reserve(keys, keys_offset);
+	ListVector::Reserve(children, children_offset);
+	ListVector::Reserve(values, values_offset);
+
+	//! Set list sizes
+	ListVector::SetListSize(keys, keys_offset);
+	ListVector::SetListSize(children, children_offset);
+	ListVector::SetListSize(values, values_offset);
+
+	keys_selvec.Initialize(keys_offset);
+	//! NOTE: we initialize these to 0 because some rows will not set them, if the row is NULL
+	//! To simplify the implementation we just allocate 'dictionary.Size()' keys for each row
+	for (idx_t i = 0; i < keys_offset; i++) {
+		keys_selvec.set_index(i, 0);
```

Comment:
I think this can be removed, it's copied from `to_variant.cpp`, but in this case we actually know that the keys are used